cpp/src/ray/runtime/task/task_executor.cc

```diff
@@ -169,23 +179,29 @@ void TaskExecutor::Invoke(
   auto typed_descriptor = function_descriptor->As<ray::CppFunctionDescriptor>();
 
   std::shared_ptr<msgpack::sbuffer> data;
-  if (actor) {
-    auto result = internal::TaskExecutionHandler(typed_descriptor->FunctionName(),
-                                                 args_buffer, actor.get());
-    data = std::make_shared<msgpack::sbuffer>(std::move(result));
-    runtime->Put(std::move(data), task_spec.ReturnId(0));
-  } else {
-    auto result = internal::TaskExecutionHandler(typed_descriptor->FunctionName(),
-                                                 args_buffer, nullptr);
-    data = std::make_shared<msgpack::sbuffer>(std::move(result));
-    if (task_spec.IsActorCreationTask()) {
-      std::unique_ptr<ActorContext> actorContext(new ActorContext());
-      actorContext->current_actor = data;
-      absl::MutexLock lock(&actor_contexts_mutex);
-      actor_contexts.emplace(task_spec.ActorCreationId(), std::move(actorContext));
-    } else {
+  try {
+    if (actor) {
+      auto result = internal::TaskExecutionHandler(typed_descriptor->FunctionName(),
+                                                   args_buffer, actor.get());
+      data = std::make_shared<msgpack::sbuffer>(std::move(result));
       runtime->Put(std::move(data), task_spec.ReturnId(0));
+    } else {
+      auto result = internal::TaskExecutionHandler(typed_descriptor->FunctionName(),
+                                                   args_buffer, nullptr);
+      data = std::make_shared<msgpack::sbuffer>(std::move(result));
+      if (task_spec.IsActorCreationTask()) {
+        std::unique_ptr<ActorContext> actorContext(new ActorContext());
+        actorContext->current_actor = data;
+        absl::MutexLock lock(&actor_contexts_mutex);
+        actor_contexts.emplace(task_spec.ActorCreationId(), std::move(actorContext));
+      } else {
+        runtime->Put(std::move(data), task_spec.ReturnId(0));
+      }
     }
+  } catch (std::exception &e) {
+    auto result = ray::internal::PackError(e.what());
+    auto data = std::make_shared<msgpack::sbuffer>(std::move(result));
+    runtime->Put(std::move(data), task_spec.ReturnId(0));
```

Comment:
Why do we use "put" here? We should only use "put" if it's a large object. Please call `AllocateReturnObject` and `SealReturnObject` instead.

BTW, It seems that `TaskExecutor::Invoke` is only used in local mode. Why?

Reply:
This runtime is local mode runtime, the put will call LocalModeObjectStore::PutRaw, and then put the data into memory_store. The native mode call AllocateReturnObject and SealReturnObject.

Invoke is for local mode, and ExecuteTask is for native mode.